llvm/lib/CodeGen/CheriBoundAllocas.cpp

```diff
@@ -295,44 +295,60 @@ class CheriBoundAllocas : public ModulePass, public InstVisitor<CheriBoundAlloca
                         << (S ? Twine(*S) : Twine("<unknown>"));
                  AI->dump());
 
-      Value *SingleBoundedAlloc = nullptr;
       if (ReuseSingleIntrinsicCall) {
+        // If we use a single instrinsic for all uses, we can simply update
+        // all uses to point at the newly inserted intrinsic.
         NumSingleIntrin++;
         // We need to convert it to an i8* for the intrinisic:
-        Instruction *AllocaI8 =
-          cast<Instruction>(B.CreateBitCast(AI, I8CapTy));
-        SingleBoundedAlloc = B.CreateCall(SetBoundsIntrin, {AllocaI8, Size});
+        Instruction *AllocaI8 = cast<Instruction>(B.CreateBitCast(AI, I8CapTy));
+        Value *SingleBoundedAlloc =
+            B.CreateCall(SetBoundsIntrin, {AllocaI8, Size});
         SingleBoundedAlloc = B.CreateBitCast(SingleBoundedAlloc, AllocaTy);
-      }
-      for (const Use *U : UsesThatNeedBounds) {
-        Instruction *I = cast<Instruction>(U->getUser());
-        if (ReuseSingleIntrinsicCall) {
+        for (const Use *U : UsesThatNeedBounds) {
           const_cast<Use *>(U)->set(SingleBoundedAlloc);
-        } else {
-          if (auto PHI = dyn_cast<PHINode>(I)) {
-            // For PHI nodes we can't insert just before the PHI, instead we
-            // must insert it just before
-            auto BB = PHI->getIncomingBlock(*U);
-            // LLVM_DEBUG(dbgs() << "PHI use coming from"; BB->dump());
-            B.SetInsertPoint(BB->getTerminator());
+        }
+      } else {
+        // Otherwise, we create new intrinsics for every use. This can avoid
```

Comment:
Hm, why do we actually bother doing this? If it's rematerialisable, shouldn't it not matter?

Reply:
Good point. I'm afraid I don't remember exactly why it needs to be done here. Doing it later should be better but there were some issues with rematerialization, so I ended up manually doing it in this pass.

Reply:
Hmmm maybe that logic predates `CheriBoundedStackPseudo`, or maybe the MIR rematerialization caused some bad codegen...